internal/server/GameLoop.go

```diff
@@ -3,30 +3,115 @@ package server
 import (
 	"SOMAS2023/internal/common/objects"
 	"SOMAS2023/internal/common/physics"
+	"SOMAS2023/internal/common/utils"
 	"fmt"
+
+	"github.com/google/uuid"
 )
 
 func (s *Server) RunGameLoop() {
-	// Each agent makes a decision
+
+	// -------------------------- PROCESS JOINING REQUESTS -------------------------
+	// 1. group agents that have onBike = false by the bike they are trying to join
+	bikeRequests := s.GetJoiningRequests()
+	// 2. pass to agents on each of the desired bikes a list of all agents trying to join
+	for bike, pendingAgents := range bikeRequests {
+		agents := s.megaBikes[bike].GetAgents()
+
+		responses := make([](map[uuid.UUID]bool), 0) // list containing all the agents' ranking
+		for _, agent := range agents {
+			responses = append(responses, agent.DecideJoining(pendingAgents))
+		}
+		// 3. accept agents based on the response outcome (it will have to be a ranking system, as only 8-n bikers can be accepted)
+		acceptedRanked := GetAcceptanceRanking(responses)
+		for _, accepted := range acceptedRanked[:(8 - len(s.megaBikes[bike].GetAgents()))] {
+			s.GetAgentMap()[accepted].ToggleOnBike()
+			s.SetBikerBike(s.GetAgentMap()[accepted], bike)
+		}
+	}
+
+	// map of the proposed lootboxes by bike (for each bike a list of lootbox proposals is made, with one lootbox proposed by each agent on the bike)
+	proposedDirections := make(map[uuid.UUID][]uuid.UUID)
+	// -------------------------------- DECIDE ACTION ---------------------------------
 	for agentId, agent := range s.GetAgentMap() {
 		fmt.Printf("Agent %s updating state \n", agentId)
 		agent.UpdateGameState(s)
 		agent.UpdateAgentInternalState()
+
 		switch agent.DecideAction() {
 		case objects.Pedal:
```

Comment:
I'd consider the following: have separate phases for this. 

1. Start with the `willChangeBike` mechanism - this calls 'ToggleBike', putting them in limbo
2. Now get a data structure of {pedal, brake, turn}, which can all be processed simultaneously by the server. If an agent has `onBike == true`, then you can take it into account, else just skip over the ones that aren't on a bike
3. Handle the bike changes - this can also be done before step two, depending on how you want the events ordered (i.e., if agents switching bikes are allowed to pedal on the same turn)

Reply:
This allows you to decompose the game loop into phases (and test them) - i.e., 

1. `RunBikeSwitch`
2. `RunActionProcess`
3. `RunLeaderElection`

or something like this